File: packages/typescript/goldenanalysis/src/core/narrative.ts
/**
 * Narrative generation — one-paragraph NL summary from the flagged regressions +
 * the largest co-moving metrics. Deterministic, ASCII-clean. Parity with the Python
 * `narrative.py`.
 */

import type { Regression } from "./regressions.js";
import type { AnalysisReport } from "./types.js";

function pretty(key: string): string {
  const last = key.split(".").pop() ?? key;
  return last.replace(/_/g, " ");
}

function capitalize(text: string): string {
  return text.length === 0 ? text : text[0]!.toUpperCase() + text.slice(1);
}

function num(value: number): string {
  return String(value);
}

function topFindingClass(report: AnalysisReport): readonly [string, number] | null {
  for (const table of report.tables) {
    if (table.name === "findings_by_class" && table.rows.length > 0) {
      let best: readonly unknown[] = table.rows[0]!;
      for (const row of table.rows) {
        const c = typeof row[1] === "number" ? row[1] : 0;
        const b = typeof best[1] === "number" ? best[1] : 0;
        if (c > b) best = row;
      }
      return [String(best[0]), Number(best[1])] as const;
    }
  }
  return null;
}

export function buildNarrative(report: AnalysisReport, regressions: readonly Regression[] = []): string {
  const flagged = regressions.filter((r) => r.flagged);

  if (flagged.length === 0) {
    const notable = report.metrics
      .filter((m) => typeof m.value === "number")
      .slice()
      .sort((a, b) => Math.abs(Number(b.value)) - Math.abs(Number(a.value)))
      .slice(0, 3);
    if (notable.length === 0) return "No metrics to summarize.";
    const bits = notable.map((m) => `${pretty(m.key)} = ${m.value}`).join(", ");
    return `No regressions flagged. Notable metrics: ${bits}.`;
  }

  let worst = flagged[0]!;
  for (const r of flagged) {
    if (Math.abs(r.deltaPct) > Math.abs(worst.deltaPct)) worst = r;
  }
  const dir = worst.deltaPct < 0 ? "fell" : "rose";
  const lead = `${capitalize(pretty(worst.metric))} ${dir} to ${num(worst.current)} (baseline ${num(worst.baseline)}; ${worst.deltaPct >= 0 ? "+" : ""}${worst.deltaPct.toFixed(1)}%).`;

  const parts = [lead];
  const comovers = flagged.filter((r) => r.metric !== worst.metric);
  if (comovers.length > 0) {
    const moves = comovers
      .map((r) => `${pretty(r.metric)} ${num(r.baseline)} -> ${num(r.current)} (${r.deltaPct >= 0 ? "+" : ""}${r.deltaPct.toFixed(1)}%)`)
      .join("; ");
    parts.push(`Co-moving signals: ${moves}.`);
  }
  const fc = topFindingClass(report);
  if (fc !== null) parts.push(`Most common quality finding: ${fc[0]} (${fc[1]}).`);
  return parts.join(" ");
}

File: packages/python/goldenanalysis/goldenanalysis/narrative.py
"""Narrative generation — a one-paragraph NL summary templated from the flagged
regressions + the largest co-moving metrics across analyzers.

Deterministic; ASCII only (Windows-terminal safe, no em-dash). The root cause in
the spec's worked scenario was only visible by crossing analyzers, so this operates
over the full metric set, not one analyzer's output.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from goldenanalysis.models import AnalysisReport, Regression


def _pretty(key: str) -> str:
    return key.split(".")[-1].replace("_", " ")


def _top_finding_class(report: AnalysisReport) -> tuple[str, int] | None:
    for table in report.tables:
        if table.name == "findings_by_class" and table.rows:
            top = max(table.rows, key=lambda r: r[1] if len(r) > 1 and isinstance(r[1], int) else 0)
            return str(top[0]), int(top[1])
    return None


def build_narrative(report: AnalysisReport, regressions: list[Regression] | None = None) -> str:
    regressions = regressions or []
    flagged = [r for r in regressions if r.flagged]

    if not flagged:
        # Neutral summary: the most notable metrics by magnitude.
        notable = sorted(
            (m for m in report.metrics if isinstance(m.value, (int, float))),
            key=lambda m: abs(float(m.value)),
            reverse=True,
        )[:3]
        if not notable:
            return "No metrics to summarize."
        bits = ", ".join(f"{_pretty(m.key)} = {m.value}" for m in notable)
        return f"No regressions flagged. Notable metrics: {bits}."

    worst = max(flagged, key=lambda r: abs(r.delta_pct))
    lead = (
        f"{_pretty(worst.metric).capitalize()} {'fell' if worst.delta_pct < 0 else 'rose'} to "
        f"{worst.current:g} (baseline {worst.baseline:g}; {worst.delta_pct:+.1f}%)."
    )

    comovers = [r for r in flagged if r.metric != worst.metric]
    parts = [lead]
    if comovers:
        moves = "; ".join(
            f"{_pretty(r.metric)} {r.baseline:g} -> {r.current:g} ({r.delta_pct:+.1f}%)" for r in comovers
        )
        parts.append(f"Co-moving signals: {moves}.")
    fc = _top_finding_class(report)
    if fc is not None:
        parts.append(f"Most common quality finding: {fc[0]} ({fc[1]}).")
    return " ".join(parts)
